nvme.c

```diff
@@ -221,70 +217,82 @@ void *nvme_realloc(void *p, size_t len)
 	return result;
 }
 
-static void *__nvme_alloc_huge(size_t len, bool *huge)
-{
+#define HUGE_MIN 0x80000
+
+struct nvme_mem_huge {
+	size_t len;
+	bool nvme_malloc; /* p has been allocated using nvme_malloc */
 	void *p;
+};
 
-	if (!posix_memalign(&p, getpagesize(), len)) {
-		*huge = false;
-		memset(p, 0, len);
-		return p;
-	}
-	return NULL;
-}
+struct nvme_mem_huge *nvme_alloc_huge(size_t len)
+{
+	struct nvme_mem_huge *mh;
 
-#define HUGE_MIN 0x80000
+	mh = malloc(sizeof(*mh));
+	if (!mh)
+		return NULL;
 
-#ifdef CONFIG_LIBHUGETLBFS
-void nvme_free_huge(void *p, bool huge)
-{
-	if (huge) {
-		if (p)
-			free_hugepage_region(p);
-	} else {
-		free(p);
+	memset(mh, 0, sizeof(*mh));
```

Comment:
Use `calloc()` instead?

Reply:
I am using `posix_memalign` here directly now to allocate at a 2MB boundary. So the memset is still necessary.

Reply:
I meant for the `nvme_mem_huge`, which is what's being zeroed here, not the data memory